specification/dartLangSpec.tex

```diff
@@ -11716,18 +11716,43 @@ \subsection{Function Expressions}
 }
 
 \LMHash{}%
-We say that $S$ is the
-\IndexCustom{future type derived from a type}{type!future type derived from}
-$T$ in the following cases, using the first applicable case:
+We say that a type $T$
+\IndexCustom{derives a future type}{type!derives a future type}
+$F$ in the following cases, using the first applicable case:
 
+%% TODO(eernst): Note that `X extends X?` can create an infinite loop.
+%% We will probably make that kind of bound an error.
 \begin{itemize}
-\item $T$ implements $S$
-  (\ref{interfaceSuperinterfaces}),
-  and there is a $U$ such that $S$ is \code{Future<$U$>}.
-\item $T$ is $S$ bounded
-  (\ref{bindingActualsToFormals}),
-  and there is a $U$ such that
-  $S$ is \code{FutureOr<$U$>}, \code{Future<$U$>?}, or \code{FutureOr<$U$>?}.
+\item
+  %% TODO(eernst): Come mixin classes and extension types: add them.
+  If $T$ is a type which is introduced by
+  a class, mixin, or enum declaration,
+  and if $T$ or a direct or indirect superinterface
+  (\ref{interfaceSuperinterfaces})
+  of $T$ is \code{Future<$U$>} for some $U$,
+  then $T$ derives the future type \code{Future<$U$>}.
+\item
+  If $T$ is the type \code{FutureOr<$U$>} for some $U$,
+  then $T$ derives the future type \code{FutureOr<$U$>}.
+\item
+  If $T$ is \code{$S$?} for some $S$, and
+  $S$ derives the future type $F$,
+  then $T$ derives the future type \code{$F$?}.
+\item
+  If $T$ is a type variable with bound $B$, and
+  $B$ derives the future type $F$,
+  then $T$ derives the future type $F$.
+\item
+  If $T$ is of the form \code{$X$\,\&\,$S$} then:
+  \begin{itemize}
+  \item
+    if $S$ derives the future type $F$,
+    then $T$ derives the future type $F$.
+  \item
+    if $B$ does not derive a future type,
+    but $X$ derives the future type $F$,
+    then $T$ derives the future type $F$.
```

Comment:
(Not sure about the last case. The way I can see it's reachable are constructions like:
* `(X extends FutureOr<int>?) & int` - the non-future branch of a `FutureOr`
* `(X extends FutureOr<int>?) & Null` - the non-future-deriving branch of a `?`
* `(X extends FutureOr<int>?) & Never` - Never, always fun, never useful.

In all of these cases, I'd actually prefer to say that they type does *not* derive a future type, since we've chosen, by type-checking to get promotion, the *non-future* branch of the type. We've *checked* that it's not a future! So let's grant the user that.)